### ISSUE TITLE
Consider necessary changes/support for Tk 9.x

### ISSUE THREAD

**Issue by tangledhelix** [DistributedProofreaders/guiguts-py#605]
### Previous title: [Mac only] Homebrew Python >= 3.12 (Tk 9.x?) rendering changes compared to 3.11

See screenshot (this is from Mac). On the left, `pip install` in a venv. On the right, my `poetry run` from dev env. Note that they are both set to the same theme and even think they are using the same font size.

Perhaps some theme files are missing from the pip install? If I choose light or dark themes it *does* switch... but these are clearly not equivalent.

<img width="1316" alt="Screenshot 2024-12-26 at 3 00 13 PM" src="https://github.com/user-attachments/assets/f208c655-0740-4f8d-a090-e6be3bbca420" />

**Comment by tangledhelix** [comment]
Dark theme. This is more consistent (e.g. cursorline seems to be the same) but there are still differences, such as the font sizes aren't the same, and some other things aren't the same, like the color of checkboxes in the settings pane.

<img width="1328" alt="Screenshot 2024-12-26 at 3 04 48 PM" src="https://github.com/user-attachments/assets/8bc4fc21-99ef-4c65-84fd-3fe5e5f2b71f" />

**Comment by windymilla** [comment]
That's very strange. The theme files are supposed to be in guiguts/data/themes/awthemes-10.4.0 and are a subset of the ones in the awthemes package.

I've just taken the alpha.12 .tar.gz distribution file and unzipped it, then done a directory comparison between the themes dir we distribute and the themes dir in my dev area - all the files are identical, with none missing.

**Comment by windymilla** [comment]
I don't see any differences between my pip installed version and my dev version on Windows, so ATM it looks like Mac-only.

**Comment by srjfoo** [comment]
And I don't see a difference between running in my dev environment and running poetry run in a different shell window. I have a very simple dev environment, though, not plumbed into VSCode.

**Comment by tangledhelix** [comment]
> And I don't see a difference between running in my dev environment and running poetry run in a different shell window. I have a very simple dev environment, though, not plumbed into VSCode.

I don't think poetry run would reproduce it. You'd need a "regular user" install. To do it cleanly without installing anything in your global Python environment, use a venv:

```
mkdir foo
cd foo
python3 -m venv venv
. venv/bin/activate
pip install guiguts
guiguts
```

**Comment by tangledhelix** [comment]
(And then compare that to a run in your normal dev env using poetry)

**Comment by tangledhelix** [comment]
For the sake of completeness, Light theme - notice that that cursorline is correct now, as opposed to the wrong cursorline using Default theme + OS light mode. This one is most confusing of all, since the coloring here doesn't even look at the configured theme name, it only uses `is_dark_theme()` which uses the text color to decide what theme we're in.

<img width="1425" alt="Screenshot 2024-12-26 at 10 46 43 PM" src="https://github.com/user-attachments/assets/5887b091-ae03-4cf4-97e3-d60ee933bb2c" />

**Comment by tangledhelix** [comment]
(And again in light theme, the checkboxes are not the same...)

**Comment by tangledhelix** [comment]
One more screenshot (on font sizes).

<img width="366" alt="Screenshot 2024-12-26 at 11 12 53 PM" src="https://github.com/user-attachments/assets/3c7c46c0-2968-4339-88cb-717b164b26fb" />

Top row: left, my `pip install` guiguts, and right, a TextEdit window with DP Sans Mono 16 text (ignore that it says 18 in the text)

Bottom row: left, my `poetry run` guiguts, and right, a TextEdit window with DP Sans Mono 12 text.

**Comment by tangledhelix** [comment]
This is probably a Mac-only issue. It seems to stem from a change between Homebrew's Python 3.11 and 3.12+, where 3.12+ is now using Tk 9.0. I'm not clear what Tk 9.0 change(s) might matter. But using Homebrew Python3.11 (which reports Tk 8.6), it works fine.

For the time being, using Python 3.11 from Homebrew works, and python.org's Python seems to work. Homebrew Python 3.12+ gives strange results.

Since this seems to be caused by a change in Homebrew, it probably doesn't affect Windows or Linux.

**Comment by windymilla** [comment]
@dlowe, what do you think about either of the following actions:
1. Edit the title of this issue to reflect that the theme differences are not to do with pip install vs poetry, but something like "look at differences in Tk9.0", then change it from "bug" to "future feature".
2. Or, just close it, since we can't fix it.

**Comment by dlowe** [comment]
> @dlowe, what do you think about either of the following actions:

Pretty sure you've tagged the wrong person :)

**Comment by windymilla** [comment]
> > @dlowe, what do you think about either of the following actions:
> 
> Pretty sure you've tagged the wrong person :)

Oops, so I did. Sorry

**Comment by tangledhelix** [comment]
> @dlowe, what do you think about either of the following actions:
> 
> 1. Edit the title of this issue to reflect that the theme differences are not to do with pip install vs poetry, but something like "look at differences in Tk9.0", then change it from "bug" to "future feature".
> 2. Or, just close it, since we can't fix it.

I don't think we should close it, because it appears to be linked to Homebrew's Python which I anticipate future users will want to to be able to use.

I also don't think it's a feature; https://github.com/DistributedProofreaders/guiguts-py/labels/code%20improvement might be a better label to use here? This is more about forward compatibility with one of our dependencies.

**Comment by tangledhelix** [comment]
> > what do you think about either of the following actions:
> > 
> > 1. Edit the title of this issue to reflect that the theme differences are not to do with pip install vs poetry, but something like "look at differences in Tk9.0", then change it from "bug" to "future feature".
> > 2. Or, just close it, since we can't fix it.
> 
> I don't think we should close it, because it appears to be linked to Homebrew's Python which I anticipate future users will want to to be able to use.
> 
> I also don't think it's a feature; [ code improvement](https://github.com/DistributedProofreaders/guiguts-py/labels/code%20improvement) Improve maintainability, consistency, etc might be a better label to use here? This is more about forward compatibility with one of our dependencies.

**Comment by windymilla** [comment]
See #689 for a more serious issue apparently relating to Tk 9.x

**Comment by windymilla** [comment]
Further to #689...

See `handle_popdown_font()` in widgets.py
It calls:
```
        popdown = self.tk.eval(f"ttk::combobox::PopdownWindow {self}")
        self.tk.call(f"{popdown}.f.l", "configure", "-font", self["font"])
```
which is the equivalent of the following [in the Tk 9 docs](https://www.tcl-lang.org/man/tcl9.0/TkCmd/ttk_combobox.html):

> To configure a specific listbox (subject to future change):
> ```
> set popdown [ttk::combobox::PopdownWindow .mycombobox]
> $popdown.f.l configure -font font
> ```


Note "subject to future change" - perhaps it changed between Tk 8 & 9, but the docs haven't been updated. 

I think Matt's error message would be explained if either the ComboboxPopdownFrame is no longer named `f`, or the Listbox within it is no longer named "l".

If the above is correct, then the solution is either to determine the name via a better algorithm (e.g. get children of the combobox and find the frame/listbox like that) or to find out what the name is in Tk9, and switch on the Tk version number.,

**Comment by windymilla** [comment]
This has had enough done for now
